protocols/v2/codec-sv2/src/decoder.rs

```diff
@@ -34,22 +71,51 @@ use crate::Error::MissingBytes;
 #[cfg(feature = "noise_sv2")]
 use crate::State;
 
-#[cfg(feature = "noise_sv2")]
-pub type StandardNoiseDecoder<T> = WithNoise<Buffer, T>;
+/// A decoded frame that could either be a regular or Noise-protected frame.
 pub type StandardEitherFrame<T> = EitherFrame<T, <Buffer as IsBuffer>::Slice>;
+
+/// A decoded Sv2 frame.
 pub type StandardSv2Frame<T> = Sv2Frame<T, <Buffer as IsBuffer>::Slice>;
+
+/// Standard decoder with Noise protocol support.
+#[cfg(feature = "noise_sv2")]
+pub type StandardNoiseDecoder<T> = WithNoise<Buffer, T>;
+
+/// Standard Sv2 decoder without Noise protocol support.
 pub type StandardDecoder<T> = WithoutNoise<Buffer, T>;
 
+/// Decoder for Sv2 frames with Noise protocol support.
 #[cfg(feature = "noise_sv2")]
 pub struct WithNoise<B: IsBuffer, T: Serialize + binary_sv2::GetSize> {
+    /// Marker for the type of frame being decoded.
+    ///
+    /// Used to maintain type information for the generic parameter `T` which represents the type
+    /// of frames being decoded. `T` refers to a type that implements the necessary traits for
+    /// serialization (`binary_sv2::Serialize`), deserialization (`binary_sv2::Deserialize`), and
+    /// size calculation (`binary_sv2::GetSize`).
     frame: PhantomData<T>,
+
+    /// Number of missing bytes needed to complete the Noise header or payload.
+    ///
+    /// Keeps track of how many more bytes are required to fully decode the current Noise frame.
     missing_noise_b: usize,
+
+    /// Buffer for holding encrypted Noise data.
+    ///
+    /// Stores the incoming encrypted data until it is ready to be decrypted and processed.
     noise_buffer: B,
+
+    /// Buffer for holding decrypted Sv2 data.
+    ///
+    /// Stores the decrypted data from the Noise protocol until it is ready to be processed and
+    /// concerted into a Sv2 frame.
     sv2_buffer: B,
 }
 
 #[cfg(feature = "noise_sv2")]
 impl<'a, T: Serialize + GetSize + Deserialize<'a>, B: IsBuffer + AeadBuffer> WithNoise<B, T> {
+    /// Attempts to decode the next frame, returning either a frame or an error indicating how many
```

Comment:
Can also return other kind of error. So it return either a frame, an error, or an error with missing bytes. I did it like that but I think that maybe here having and enum into the Ok and the error would have been a better interface. Not sure if make sense to change it now.

Reply:
also here an example would be nice. I just stop writing this but I think that when is possible we should add examples

Reply:
Added `examples/` and documentation surrounding the `MissingBytes` error situation.